Added option to allow implementing custom discriminator with DiscriminatorDefiner metadata option

kraendavid opened the pull request:
Added option to allow implementing custom discriminator with custom RDBMSStoreManager.
Also made it easier to subclass RDBMSStoreManager + made performance optimization.

andyjefferson commented:
Tests should be named based on what they test. ColumnReuseTest made sense (because it was testing reuse of columns). TransportationTest doesn't (because DN doesn't have a feature called that). 

Please define (text) what this "custom discriminator" does that is not achievable using standard JDO mechanisms

kraendavid commented:
Just updated pull request with more comments and re-introduced ColumnReuseTest and renamed TransportationTest to CustomDiscriminatorTest.

From comment in code:
The custom class-name-resolver enables to discriminate persistent
objects from the selected ResultSet - eg. by using more column values
to make a decision on which class should be instantiated from this DB row.
This calculation can be of any complexity using the DB row from ResultSet.
Current implementation only allows for discriminating on one string column
which might be enough in many cases - but this custom mechanism enables any
complexity for resolving persistent classes based on a DB row.
